### PR TITLE
Separate Paste Handler

### LINKED ISSUE #6102: Classic + Custom HTML blocks: Convert to Blocks removes valid inline formatting
## Issue Overview
The Paragraph block, as well as other textual blocks, allow `<abbr>`, `<b>`, `<code>`, `<i>`, `<kbd>`, `<mark>`, `<span>`, `<time>`, and various other inline formatting and semantic tags to be added using the "Edit as HTML" option, and these are considered valid and are not removed on save.

However, when converting a Classic block to standard blocks using the "Convert to Blocks" option, the paragraphs, lists, and blockquotes are stripped of some inline formatting tags, and a few are converted to tags that have a different semantic meaning.

I think inline formatting that is considered valid by the standard blocks should not be removed when converting a Classic block to standard blocks. Otherwise, you are stripping out formatting from the original post when you do not need to.

## Steps to Reproduce (for bugs)
1. Create a post using the Classic Editor or insert a Classic block in the Gutenberg editor.
2. Insert the following:
```html
<abbr>abbr</abbr> <b>b</b> <br>br <bdi>bdi</bdi> <bdo dir="rtl">bdo</bdo> <cite>cite</cite> <code>code</code> <data value="value">data</data> <dfn>dfn</dfn> <em>em</em> <i>i</i> <kbd>kbd</kbd> <mark>mark</mark> <q>q</q> <ruby>ruby <rb>rb</rb> <rp>rp</rp> <rt>rt</rt> <rtc>rtc</rtc> <rp>rp</rp></ruby> <s>s</s> <samp>samp</samp> <small>small</small> <span style="color:red">span</span> <strong>strong</strong> <sub>sub</sub> <sup>sup</sup> <time datetime="2018">time</time> <u>u</u> <var>var</var> <wbr>wbr
```
3. Save the post.
4. Open the post in the Gutenberg editor.
5. Convert the Classic block to standard blocks using the <kbd><samp>Convert to Blocks</samp></kbd> option.
6. Notice how some of the inline formatting has been removed, and some of it has been converted to other elements with different semantic meaning, e.g. `<b>` tags are converted to `<strong>` tags.

## Expected Behavior
Converting a Classic block to standard blocks using the "Convert to Blocks" option should preserve all inline formatting tags that are considered valid by the resulting blocks.

## Current Behavior
Converting a Classic block to standard blocks using the "Convert to Blocks" option removes several valid HTML5 tags, and converts some tags to other tags with different semantic meanings, e.g. `<b>` tags are converted to `<strong>` tags. The previously given sample input above is transformed into the following:
```html
<p><abbr>abbr</abbr> <strong>b</strong> <br/>br bdi bdo cite <code>code</code> data dfn <em>em</em> <em>i</em> kbd mark q ruby rb rp rt rtc rp s samp small span <strong>strong</strong> <sub>sub</sub> <sup>sup</sup> time u var wbr</p>
```

## Other Notes
If a Classic block contains obsolete HTML tags like `<color>`, then converting them to `<span>` tags upon conversion to standard blocks seems like a good idea.

`<b>` and `<i>` tags are converted to `<strong>` and `<em>` respectively. However, it should be considered that there are valid cases where a `<b>` or `<i>` are used semantically as per the HTML5 specification. Additionally, `<span>` tags containing `font-weight:bold` or `font-style:italic` are converted to `<strong>` or `<em>` tags respectively, and this seems like a bad idea. Most of the time, when someone makes text bold or italic using a `<span>` tag, they are doing it for purely stylistic reasons and not semantic reasons, so converting those to `<strong>` and `<em>` tags seems like a bad idea.

## Related Issues and/or PRs
- #4186
- <s>#4498</s>
- <s>#5876</s>
- <s>#6096</s>
- #6878
- #7604

### PR DESCRIPTION
**Pull request by ellatrix**
## Description
Fixes #6102.

This PR separates paste handling from raw handling in general. I opted of a separate function instead of an extra argument because raw handling is significantly less complex and doesn't need as many arguments to works with. Raw handling aims to convert HTML without comment delimiters to blocks, trying to retain as much content as possible.

Needs an integration test.

## How has this been tested?
Convert HTML/Classic block to blocks with e.g. a span with attributes in a paragraph.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation.

### CODE REVIEW

**Comment by danielbachhuber** [comment]
@iseulde I tried testing with `<p>This classic block has <code>a code snippet</code>.</p>` but got the same behavior on this branch and `master`:

convertinline

Can you update your testing instructions to include more detail about how to reproduce the behavior of the change? Thanks

**Comment by iseulde** [comment]
@danielbachhuber Thanks, updated. The code tag will also be preserved on paste. Try something that wouldn't e.g. a `span` with attributes or a `small` tag.

**Review by danielbachhuber — COMMENTED** [review]
Confirmed that the `<small>` tag persists when converting from Classic to Paragraph Block:

convertinline

I'll defer to others on the underlying implementation.

File: packages/blocks/src/api/raw-handling/index.js [hunk 1]
```diff
@@ -225,3 +229,41 @@ export default function rawHandler( { HTML = '', plainText = '', mode = 'AUTO',
...
+	const rawTransformations = getRawTransformations();
+	const blockContentSchema = getBlockContentSchema( rawTransformations );
+
+	return compact( flatMap( pieces, ( piece ) => {
+		// Already a block from shortcode.
+		if ( typeof piece !== 'string' ) {
+			return piece;
+		}
+
+		const filters = [
```
Comment (aduth) [inline packages/blocks/src/api/raw-handling/index.js]: How do we decide which filters are to be included here when one of the main points of distinguishing raw / paste is the strictness of its filtering?

File: packages/blocks/src/api/raw-handling/index.js [hunk 2]
```diff
@@ -65,6 +65,40 @@ function getRawTransformations() {
 		} );
 }
 
+function htmlToBlocks( { html, rawTransformations } ) {
```
Comment (aduth) [inline packages/blocks/src/api/raw-handling/index.js]: Could do for a JSDoc.

**Review by aduth — APPROVED** [review]
I'm not totally convinced that paste and raw are fundamentally separate things, vs. the former applying a more strict set of filters in the general operation of raw transformation. It makes me wonder if `filters` should be handled as an argument of the HTML -> blocks transform function. Then again, this still means we'd have to either decide between exposing filters on the public interface, or otherwise provide just as we have done here with an abstracted form ("paste", "raw"). Maybe then this consideration just becomes a future refactoring detail.

Overall, I think this is a positive change 👍

**Comment by susanpaigen** [comment]
If this is the wrong place to post this I apologize. I've been posting to the #6012 string and I'm not sure whether to comment here or there.
Does this solution preserve attributes as well as tags? e.g. class="".
Does it also preserve attributes in other types of content, e.g. headings or lists?

Testing what happens now with the Gutenberg plugin:
I make a paragraph or list block, add content, 'edit in html', add markup (a span with a class, a style on a span or <li>).
If I copy the marked-up content in 'edit in html', make a new block, switch to 'edit in html' in the new block, paste my marked-up content, and update, the markup is preserved.
If I copy the marked-up content in 'edit visually', make a new block, switch to 'edit in html' in the new block, paste my copied (marked-up) content, and update, the markup is stripped.
So there's some kind of existing copy/paste functionality which doesn't strip the markup.

**Comment by danielbachhuber** [comment]
Hi @susanpaigen!

> Does this solution preserve attributes as well as tags? e.g. class="".
> Does it also preserve attributes in other types of content, e.g. headings or lists?

This pull request doesn't, no. There's an ongoing conversation in #11440 on to what degree Gutenberg will support those behaviors. If you'd like, please do weigh in on #11440 with more details of your use-case, the markup you'd like to preserve, etc. Thanks :)

File: packages/blocks/src/api/raw-handling/index.js (hunk 1 above)
Comment (iseulde) [reply packages/blocks/src/api/raw-handling/index.js]: I'll add a comment. These are the only filters that are really needed for us to be able to transform the HTML into blocks. One is for the read more tag, one is to ensure images etc are properly positioned, and the last is to ensure text content in quotes without a wrapper tag are wrapped in a paragraph. None of these remove any content.

**Pull request merged**